### PR TITLE
IEP-352 Add choice whether to flash application over JTAG or UART

### PR DESCRIPTION
**Pull request by sigmaaa**
Added option to flash application over JTAG. This option is available only with the latest [openocd v0.10.0-esp32-20210401](https://github.com/espressif/openocd-esp32/releases/tag/v0.10.0-esp32-20210401).

basic features:

- "Flash over JTAG" option in the project main configuration tab is available for a user only if he has openocd v0.10.0-esp32-20210401
- when "Flash over JTAG" is selected, only the UI, which is responsible for jtag flash will be enabled (and vice versa):

Screenshot 2021-06-14 at 9 31 38

- Users can easily build the arguments string based on their selection in fields "Flash voltage", "target, "board". 

BAO: when selecting ESP-WROVER-KIT 1.8V or ESP-WROVER-KIT 3.3V path to config file is "boards/*" instead of "board". 
so for now, the user has to edit this path in the arguments field.

### CODE REVIEW

File: bundles/com.espressif.idf.launch.serial.core/src/com/espressif/idf/launch/serial/internal/SerialFlashLaunchConfigDelegate.java [hunk 1]
```diff
@@ -150,6 +164,34 @@ protected void launchInternal(ILaunchConfiguration configuration, String mode, I
 		DebugPlugin.newProcess(launch, p, String.join(" ", commands)); //$NON-NLS-1$
 	}
 
+	private void flashOverJtag(ILaunchConfiguration configuration, ILaunch launch) throws CoreException {
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.launch.serial.core/src/com/espressif/idf/launch/serial/internal/SerialFlashLaunchConfigDelegate.java:167]: Having the method modifier "protected" would help us to write a JUnit test case later.

File: bundles/com.espressif.idf.launch.serial.core/src/com/espressif/idf/launch/serial/internal/SerialFlashLaunchConfigDelegate.java [hunk 2]
```diff
@@ -150,6 +164,34 @@ protected void launchInternal(ILaunchConfiguration configuration, String mode, I
...
+		List<String> commands = new ArrayList<>();
+		String openocdExe = configuration.getAttribute(SERVER_EXECUTABLE, DEFAULT_PATH + DEFAULT_EXECUTABLE);
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.launch.serial.core/src/com/espressif/idf/launch/serial/internal/SerialFlashLaunchConfigDelegate.java:169]: DEFAULT_EXECUTABLE might be different on windows e.g openocd.exe, please check

File: bundles/com.espressif.idf.launch.serial.ui/src/com/espressif/idf/launch/serial/ui/internal/CMakeMainTab2.java [hunk 3]
```diff
@@ -0,0 +1,282 @@
+/*******************************************************************************
+ * Copyright 2020 Espressif Systems (Shanghai) PTE LTD. All rights reserved.
+ * Use is subject to license terms.
+ *******************************************************************************/
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.launch.serial.ui/src/com/espressif/idf/launch/serial/ui/internal/CMakeMainTab2.java:4]: As this is a copied class, please keep the original license and add espressif as a modifier

File: bundles/com.espressif.idf.launch.serial.ui/src/com/espressif/idf/launch/serial/ui/internal/CMakeMainTab2.java [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+		fileLocationButton.setEnabled(!isFlashOverJtag);
+		workspaceLocationButton.setEnabled(!isFlashOverJtag);
+		variablesLocationButton.setEnabled(!isFlashOverJtag);
+		fileWorkingDirectoryButton.setEnabled(!isFlashOverJtag);
+		workspaceWorkingDirectoryButton.setEnabled(!isFlashOverJtag);
+		variablesWorkingDirectoryButton.setEnabled(!isFlashOverJtag);
+
+		fFlashVoltage.setEnabled(isFlashOverJtag);
+		fTargetName.setEnabled(isFlashOverJtag);
+		fTarget.setEnabled(isFlashOverJtag);
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.launch.serial.ui/src/com/espressif/idf/launch/serial/ui/internal/CMakeMainTab2.java:97]: If none of these are needed for JTAG flashing, probably we can move JTAG flashing to the top and show the required UI based on the selection.

File: bundles/com.espressif.idf.ui/META-INF/MANIFEST.MF [hunk 5]
```diff
@@ -33,8 +33,23 @@ Require-Bundle: org.eclipse.core.runtime,
...
+   org.eclipse.core.resources,
+   org.eclipse.debug.ui,
+   org.eclipse.ui.plugin",
+ com.espressif.idf.ui.dialogs;
+  uses:="org.eclipse.osgi.util,
+   org.eclipse.swt.graphics,
+   org.eclipse.swt.widgets,
+   org.eclipse.jface.dialogs,
+   org.eclipse.cdt.launch.ui.corebuild,
+   org.eclipse.debug.core"
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.ui/META-INF/MANIFEST.MF:51]: Do we need these exported packages?

**Review by kolipakakondal — CHANGES_REQUESTED** [review]
This is a good PR @sigmaaa, changes looks good to me. However, can we change the UI a little bit? Will connect over a call to discuss this.

File: bundles/com.espressif.idf.launch.serial.core/src/com/espressif/idf/launch/serial/internal/SerialFlashLaunchConfigDelegate.java (hunk 1 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.launch.serial.core/src/com/espressif/idf/launch/serial/internal/SerialFlashLaunchConfigDelegate.java]: thanks for the review @kolipakakondal
changed to protected

File: bundles/com.espressif.idf.launch.serial.core/src/com/espressif/idf/launch/serial/internal/SerialFlashLaunchConfigDelegate.java (hunk 2 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.launch.serial.core/src/com/espressif/idf/launch/serial/internal/SerialFlashLaunchConfigDelegate.java:169]: @kolipakakondal 

Checked on Windows 10, execution of opencocd works

File: bundles/com.espressif.idf.launch.serial.ui/src/com/espressif/idf/launch/serial/ui/internal/CMakeMainTab2.java (hunk 3 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.launch.serial.ui/src/com/espressif/idf/launch/serial/ui/internal/CMakeMainTab2.java:13]: @kolipakakondal changed

File: bundles/com.espressif.idf.launch.serial.ui/src/com/espressif/idf/launch/serial/ui/internal/CMakeMainTab2.java (hunk 4 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.launch.serial.ui/src/com/espressif/idf/launch/serial/ui/internal/CMakeMainTab2.java]: @kolipakakondal 
good point. UI is changed

File: bundles/com.espressif.idf.ui/META-INF/MANIFEST.MF (hunk 5 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.ui/META-INF/MANIFEST.MF]: @kolipakakondal fixed

File: bundles/com.espressif.idf.core/src/com/espressif/idf/core/build/IDFLaunchConstants.java [hunk 6]
```diff
@@ -6,4 +6,5 @@
 	public static final String ATTR_IDF_TARGET =  "com.espressif.idf.launch.serial.core.idfTarget"; //$NON-NLS-1$
 	public static final String DEBUG_LAUNCH_CONFIG_TYPE= "com.espressif.idf.debug.gdbjtag.openocd.launchConfigurationType"; //$NON-NLS-1$
 	public static final String RUN_LAUNCH_CONFIG_TYPE = "com.espressif.idf.launch.serial.launchConfigurationType"; //$NON-NLS-1$
+	public static final String FLASH_OVER_JTAG = "FLASH_OVER_JTAG";
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.core/src/com/espressif/idf/core/build/IDFLaunchConstants.java:9]: Add NON-NLS

File: bundles/com.espressif.idf.launch.serial.core/src/com/espressif/idf/launch/serial/util/EspFlashCommandGenerator.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+		commands.add(IDFConstants.FLASH_CMD);
+
+		return String.join(" ", commands); //$NON-NLS-1$
+	}
+
+	private static String getLastUsedSerialPort() {
+		Preferences preferences = InstanceScope.INSTANCE.getNode("com.espressif.idf.launch.serial.ui"); //$NON-NLS-1$
+		return preferences.get("com.espressif.idf.launch.serial.core.serialPort", ""); //$NON-NLS-1$ //$NON-NLS-2$
+	}
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.launch.serial.core/src/com/espressif/idf/launch/serial/util/EspFlashCommandGenerator.java:39]: Use SerialFlashLaunchTargetProvider.ATTR_SERIAL_PORT

**Review by kolipakakondal — APPROVED** [review]
LGTM

File: bundles/com.espressif.idf.core/src/com/espressif/idf/core/build/IDFLaunchConstants.java (hunk 6 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.core/src/com/espressif/idf/core/build/IDFLaunchConstants.java]: added

File: bundles/com.espressif.idf.launch.serial.core/src/com/espressif/idf/launch/serial/util/EspFlashCommandGenerator.java (hunk 7 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.launch.serial.core/src/com/espressif/idf/launch/serial/util/EspFlashCommandGenerator.java]: @kolipakakondal new commit with requested changes:
https://github.com/espressif/idf-eclipse-plugin/pull/294/commits/e5a519a9b2433b2558df586078d9897a46545f92

Now EspFlashCommandGenerator takes serialPort string, which is taken from SerialFlashLaunchTargetProvider.ATTR_SERIAL_PORT attribute

**Pull request merged**